common/lib/util/defaults.js

```diff
@@ -102,6 +102,32 @@ Defaults.normaliseOptions = function(options) {
 		options.queueMessages = options.queueEvents;
 	}
 
+	if(options.fallbackHostsUseDefault) {
+		/* fallbackHostsUseDefault and fallbackHosts are mutually exclusive as per TO3k7 */
+		if(options.fallbackHosts) {
+			var msg = 'fallbackHosts and fallbackHostsUseDefault cannot both be set';
+			Logger.logAction(Logger.LOG_ERROR, 'Defaults.normaliseOptions', msg);
+			throw new ErrorInfo(msg, 40000, 400);
+		}
+
+		/* default fallbacks can't be used with custom ports */
+		if(options.port || options.tlsPort) {
+			var msg = 'fallbackHostsUseDefault cannot be set when port or tlsPort are set';
+			Logger.logAction(Logger.LOG_ERROR, 'Defaults.normaliseOptions', msg);
+			throw new ErrorInfo(msg, 40000, 400);
+		}
+
+		/* emit an appropriate deprecation warning */
+		if(options.environment) {
+			Logger.deprecatedWithMsg('fallbackHostsUseDefault', 'There is no longer a need to set this when the environment option is also set since the library will now generate the correct fallback hosts using the environment option.');
```

Comment:
@lmars strongly referring to this code for .net implementation.
 Not sure if this is a bug, but based on the warning, it's supposed to set env fallbacks if env is set right (even if fallbackHostsUseDefault) is used? currently, it won't be able to do that.

Reply:
> it's supposed to set env fallbacks if env is set right (even if fallbackHostsUseDefault) is used?

No, as per [RSC15g4](https://www.ably.io/documentation/client-lib-development-guide/features#RSC15g4), `fallbackHostsUseDefault` takes precedence over `environment`.

To provide some context as to why, the intention of `fallbackHostsUseDefault` is that default fallbacks can be used by customers even when setting `environment` which would otherwise disable fallback behaviour for them unless they explicitly set `fallbackHosts`, but that is no longer needed as every valid `environment` always now has associated environment fallbacks that point at the correct place (e.g. some of them are just CNAMEs to the default fallbacks), so we will eventually remove `fallbackHostsUseDefault` but for now it is just deprecated whilst keeping it's meaning (i.e. use default fallbacks even when `environment` is set).

Reply:
What about the message? I am sure, it conveys otherwise 😕

Reply:
Or maybe it's a little confusing for me

Reply:
Here's what the message says when printed:

> 22:07:14.433 Ably: Deprecation warning - 'fallbackHostsUseDefault' is deprecated and will be removed from a future version. There is no longer a need to set this when the environment option is also set since the library will now generate the correct fallback hosts using the environment option.

If it's not clear, then let's make it clearer, what do you suggest?

Reply:
Warning should mention that environment value will be overridden because of fallbackHostsUseDefault

Reply:
or maybe a little implicit message

Reply:
So let's think about the user experience here of a user that is currently setting `fallbackHostsUseDefault`.

Prior to the change that introduced environment fallbacks, let's assume the user was using the following options:

```js
var client = new Ably.Realtime({
  environment:             'acme',
  fallbackHostsUseDefault: true,
})
```

In this case, they had `fallbackHostsUseDefault` set because they wanted to use the default fallbacks. If they didn't set it, they wouldn't get any fallback behaviour at all, because the library didn't have the concept of environment fallbacks.

Now, with the recent changes that introduce environment fallbacks and deprecate `fallbackHostsUseDefault`, what should the behaviour be? Well, given we're just talking about the option being deprecated, the behaviour should be the same, they should continue to use the default fallbacks like they've been using so far, but they should be made aware that `fallbackHostsUseDefault` is deprecated and should no longer be set since now the library will transparently determine the fallback hosts from the environment.

So, to your point:

> Warning should mention that environment value will be overridden because of fallbackHostsUseDefault. To the point !!

I actually think that would be more confusing to the user, because what is being overridden? From their point of view, setting `environment` alone meant getting no fallback behaviour, so they were "overriding" that by setting `fallbackHostsUseDefault` to force the library to use some fallbacks. They no longer need to do that, and that's what I tried to convey in the deprecation message. Do you see what I mean?

I'll conclude by saying, this is really inconsequential, because hardly anyone uses the `fallbackHostsUseDefault` option, and when we remove it, whoever does use it will just start using the generated environment fallbacks and be none the wiser.

Reply:
I agree with everything lewis said ^